src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -458,7 +456,28 @@ protected virtual void CompleteTransaction(bool isCommitted)
 				{
 					// Allow transaction completed actions to run while others stay blocked.
 					_bypassLock.Value = true;
-					using (_session.BeginContext())
+					// Ensure no other session processing is still ongoing. In case of a transaction timeout, the transaction is
+					// cancelled on a new thread even for non-distributed scopes. So, the session could be doing some processing,
+					// and will not be interrupted until attempting some usage of the connection. See #3355.
+					// Thread safety of a concurrent session BeginProcess is ensured by the Wait performed by BeginProcess.
+					var context = _session.BeginProcess();
+					if (context == null)
+					{
+						var timeOutGuard = new Stopwatch();
+						timeOutGuard.Start();
+						while (context == null && timeOutGuard.ElapsedMilliseconds < _systemTransactionCompletionLockTimeout)
+						{
+							// Na�ve yield.
+							Thread.Sleep(10);
```

Comment:
Can we do this somehow without spin lock?

Reply:
I guess we could, if we were putting thread locking mechanisms directly in the session, probably whatever the transaction mechanism, instead of keeping them in the transaction factory and applied only on transaction scope finalization.

The `Wait` done by `BeginProcess` actually belongs to the transaction factory and does nothing by default: it locks only during transaction scope finalization. This was done that way to minimize the impact of locks required to handle the transaction scope end processing occurring on concurrent threads.

As a result, when the session initiates some processing inside an ongoing transaction scope, no locks are applied. So, when a transaction timeout occurs concurrently to the session processing, we have no locking mechanism enabled against which we could wait for the end of the session processing.